File: strategy_pattern/go/duck/duck.go
package duck

import "duckapp/behavior"

type Duck struct {
	flyBehavior   behavior.FlyBehavior
	quackBehavior behavior.QuackBehavior
}

func (d *Duck) PerformFly() {
	d.flyBehavior.Fly()
}

func (d *Duck) PerformQuack() {
	d.quackBehavior.Quack()
}

func (d *Duck) SetFlyBehavior(fb behavior.FlyBehavior) {
	d.flyBehavior = fb
}

func (d *Duck) SetQuackBehavior(qb behavior.QuackBehavior) {
	d.quackBehavior = qb
}

func (d *Duck) Swim() {
	println("All ducks float, even decoys!")
}

func (d *Duck) Display() {
	println("I'm a duck!")
}

File: strategy_pattern/ruby/ducks/duck.rb
class Duck
  attr_writer :fly_behavior, :quack_behavior

  def initialize
    @fly_behavior = nil
    @quack_behavior = nil
  end

  def perform_fly
    @fly_behavior&.fly
  end

  def perform_quack
    @quack_behavior&.quack
  end

  def swim
    puts 'All ducks float, even decoys!'
  end

  def display
    raise NotImplementedError, "#{self.class} must implement abstract method 'display'"
  end
end
